src/pipelinedefinition.cpp

```diff
@@ -0,0 +1,329 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "pipelinedefinition.hpp"
+
+#include <set>
+
+#include "prediction_service_utils.hpp"
+
+namespace ovms {
+
+Status toNodeKind(const std::string& str, NodeKind& nodeKind) {
+    if (str == DL_NODE_CONFIG_TYPE) {
+        nodeKind = NodeKind::DL;
+        return StatusCode::OK;
+    }
+    SPDLOG_ERROR("Unsupported node type:{}", str);
+    return StatusCode::PIPELINE_NODE_WRONG_KIND_CONFIGURATION;
+}
+
+Status PipelineDefinition::create(std::unique_ptr<Pipeline>& pipeline,
+    const tensorflow::serving::PredictRequest* request,
+    tensorflow::serving::PredictResponse* response,
+    ModelManager& manager) const {
+    std::unordered_map<std::string, std::unique_ptr<Node>> nodes;
+
+    EntryNode* entry = nullptr;
+    ExitNode* exit = nullptr;
+    for (const auto& info : nodeInfos) {
+        SPDLOG_DEBUG("Creating pipeline:{}. Adding nodeName:{}, modelName:{}",
+            getName(), info.nodeName, info.modelName);
+        switch (info.kind) {
+        case NodeKind::ENTRY: {
+            auto node = std::make_unique<EntryNode>(request);
+            entry = node.get();
+            nodes.insert(std::make_pair(info.nodeName, std::move(node)));
+            break;
+        }
+        case NodeKind::DL:
+            nodes.insert(std::make_pair(info.nodeName, std::move(std::make_unique<DLNode>(info.nodeName,
+                                                           info.modelName,
+                                                           info.modelVersion,
+                                                           manager,
+                                                           info.outputNameAliases))));
+            break;
+        case NodeKind::EXIT: {
+            auto node = std::make_unique<ExitNode>(response);
+            exit = node.get();
+            nodes.insert(std::make_pair(info.nodeName, std::move(node)));
+            break;
+        }
+        default:
+            throw std::invalid_argument("unknown node kind");
+        }
+    }
+    for (const auto& kv : connections) {
+        const auto& dependantNode = nodes.at(kv.first);
+        for (const auto& pair : kv.second) {
+            const auto& dependencyNode = nodes.at(pair.first);
+            SPDLOG_DEBUG("Connecting pipeline:{}, from:{}, to:{}", getName(), dependencyNode->getName(), dependantNode->getName());
+            Pipeline::connect(*dependencyNode, *dependantNode, pair.second);
+        }
+    }
+    pipeline = std::make_unique<Pipeline>(*entry, *exit, pipelineName);
+    for (auto& kv : nodes) {
+        pipeline->push(std::move(kv.second));
+    }
+    return StatusCode::OK;
+}
+
+void PipelineDefinition::resetSubscriptions(ModelManager& manager) {
```

Comment:
When will this method be called? There's no check if model/model instance is valid (not a nullptr). 
Are we sure values in `subscriptions` map will be valid during execution of that method?

Reply:
Yes -> we create subscriptions list & we add those models & modelInstances only if they exist. 
OVMS never removes Model nor ModelInstance so we are safe here.

You can check how it will be used here:
https://github.com/openvinotoolkit/model_server/pull/335/files#diff-ddd25336c35406ea4f13e3fc1f72437f22c39357fe89feec795c954e0aabea5bR50